Rewriting Awkward Array's GPU kernels in Python with NVIDIA's cuda.compute (#272)
scientific-python/blog.scientific-python.org: content/posts/awkward/rewriting-gpu-kernels-cuda-compute/index.md

## 1. The GPU code is Python now

`ak.min`, the minimum over each ragged sublist, took three kernel launches in CUDA C++: initialize a scratch buffer, reduce within each block using shared memory and explicit thread synchronization, and copy the result out. Each communicated with the next through global memory.

It is now a single call to a library primitive:

```python
import cupy as cp, numpy as np
from cuda.compute import OpKind, segmented_reduce

def awkward_reduce_min(
    toptr: cp.ndarray,
    fromptr: cp.ndarray,
    offsets: cp.ndarray,
    outlength: int,
    identity: float,
) -> None:

    toptr[:outlength] = identity
    segmented_reduce(
        fromptr,
        toptr,
        offsets[:-1],
        offsets[1:],
        OpKind.MINIMUM,
        np.asarray(identity, dtype=fromptr.dtype),
        outlength,
    )
```

`segmented_reduce` performs one reduction per segment, and `offsets[:-1]` and `offsets[1:]` hand it the start and end of each sublist directly. No CUDA C++, no synchronization, no scratch buffers: the library handles all of it.

> Awkward's user-facing API is unchanged. Underneath, 2.10.0 routes 106 of its 133 GPU kernels (80%) through cuda.compute, up from none in 2.8.11. Every reduction and the sort now run through it with no hand-written implementation remaining. The 27 still in CUDA C++ handle structural work like jagged indexing, padding, and validity checks. Their control flow depends on the ragged layout itself, which fits the segmented primitives less naturally. That migration is ongoing. Counting only code that must actually be maintained, hand-written CUDA C++ fell from 8,288 lines to 2,170, a 74% reduction.

Reviewer: It took me a second to grok the last sentence, maybe something like:

```suggestion
Awkward's user-facing API is unchanged. Underneath, 2.10.0 routes 106 of its 133 GPU kernels (80%) through cuda.compute, up from none in 2.8.11. Every reduction and the sort now run through it with no hand-written implementation remaining. The 27 still in CUDA C++ handle structural work like jagged indexing, padding, and validity checks. Their control flow depends on the ragged layout itself, which fits the segmented primitives less naturally. The switch to `cuda.compute` has (so far) reduced the hand-written CUDA C++ from 8,288 to 2,170, a 74% reduction.
```

Entirely subjective though - feel free to ignore!

Author: Thanks! I tried to make it a bit more clear.

> ## 2. The abstraction made it faster

Reviewer: Another purely subjective point: my takeaway here is the value of depending on well-maintained, highly-optimized upstream code!

Author: Yes, fair point! The section 2 already makes exactly that point. Thanks.

Over 5,000,000 ragged sublists, `ak.argmin` takes 0.96 ms per call as a hand-written CUDA kernel and 0.39 ms through cuda.compute: 2.5x faster, with identical output.

That speedup is inherited rather than hand-tuned, which is exactly the point of building on CUB: segmented reductions are difficult to write well by hand, and CUB's have been tuned per architecture for years. Awkward gets that tuning now, and the next architecture's when it ships, without changing its own code.

## 3. A whole physics formula can collapse into one kernel

The first two results came from the migration. For this one, a user drops down to cuda.compute and composes the primitives by hand.

Awkward evaluates eagerly: every operation returns a real array, so a chain of them writes an intermediate to global memory at each step and reads it back at the next. cuda.compute algorithms instead accept iterators that are evaluated lazily as the algorithm runs, letting many logical steps ride along inside a single pass. That is [kernel fusion](https://developer.nvidia.com/blog/kernel-fusion-in-nvidia-cuda-optimizing-memory-traffic-and-launch-overhead/), and it saves both the memory traffic and the launch overhead.

> 

Reviewer: Consider adding an L3 heading here, like `### Example: di-muon invariant mass`.

Author: done! Thanks.

| ------------------- | --------------- | ----------------- | -------- |
| step by step        | 88              | 212               | 25.7 ms  |
| fused into one call | 1               | 45                | 10.2 ms  |

2.5x faster, with nothing allocated in between.

## 4. At analysis scale, the gap widens

We ran the [ADL benchmark queries](https://github.com/CoffeaTeam/coffea-benchmarks) (a standard set of physics-analysis tasks) on CMS 2012 open data, against released Awkward 2.8.11 on its hand-written CuPy backend, the last version before cuda.compute. For the two combinatoric queries below, the Awkward expression is unchanged; only the backend underneath differs, so the comparison isolates the migration itself.

Measured on the GPU compute stage at 100k, 1M, and 10M events, both speed up by margins that grow with the data:

- the di-muon reconstruction from section 3: **60x → 422x → 3634x**
- a second combinatoric query: **45x rising to 250x**

cuda.compute's time stays approximately constant across those sizes while the hand-written implementation grows super-linearly, so the gap widens as data grows rather than closing.

## The result that isn't a number

> A stated goal of the Awkward Array project is to let physics and data analysts write high-performance code in Python without GPU expertise. The old backend required contributors to understand CUDA thread hierarchies, atomics, and shared-memory behavior before they could add or fix a kernel. The new one asks for an ordinary Python function and a call to the right primitive. Domain scientists can read it, review it, and unit-test its logic without a GPU.

Reviewer: Maybe "physicists" instead of the implied "physics analysists"? Just a suggestion - go with whatever makes the most sense to you!

```suggestion
A stated goal of the Awkward Array project is to let physicists and data analysts write high-performance code in Python without GPU expertise. The old backend required contributors to understand CUDA thread hierarchies, atomics, and shared-memory behavior before they could add or fix a kernel. The new one asks for an ordinary Python function and a call to the right primitive. Domain scientists can read it, review it, and unit-test its logic without a GPU.
```
